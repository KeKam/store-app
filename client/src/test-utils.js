import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
