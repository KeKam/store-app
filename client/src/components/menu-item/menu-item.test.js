import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import MenuItem from './menu-item';

const history = createMemoryHistory();

describe('MenuItem component', () => {
  let mockItem;

  beforeEach(() => {
    mockItem = {
      title: 'SUMMER',
      imageUrl: 'www.testImage.com',
      id: 1,
      linkUrl: 'collection/summer',
    };
  });

  it('should render MenuItem component', () => {
    const { asFragment } = render(
      <Router history={history}>
        <MenuItem {...mockItem} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent(/SUMMER/);
    expect(screen.getByText(/SEE COLLECTION/)).toBeInTheDocument();
  });

  it('should trigger history push on click', () => {
    render(
      <Router history={history}>
        <MenuItem {...mockItem} />
      </Router>
    );

    history.push = jest.fn();

    userEvent.click(screen.getByTestId('item'));

    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('/collection/summer');
  });
});
