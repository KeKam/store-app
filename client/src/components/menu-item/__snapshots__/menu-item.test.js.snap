// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem component should render MenuItem component 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa hJLFdt"
    data-testid="item"
  >
    <div
      class="sc-bwzfXH klmmtG"
      data-testid="image"
    />
    <div
      class="sc-htpNat fMwJqu"
    >
      <div
        class="sc-bxivhb hQYITJ"
        data-testid="title"
      >
        SUMMER
      </div>
      <div
        class="sc-ifAKCX jbyhsX"
      >
        SEE COLLECTION
      </div>
    </div>
  </div>
</DocumentFragment>
`;
